File: src/main/java/cube8540/oauth/authentication/credentials/oauth/token/application/ResourceOwnerPasswordTokenGranter.java
package cube8540.oauth.authentication.credentials.oauth.token.application;

import cube8540.oauth.authentication.credentials.oauth.OAuth2TokenRequest;
import cube8540.oauth.authentication.credentials.oauth.client.OAuth2ClientDetails;
import cube8540.oauth.authentication.credentials.oauth.client.domain.OAuth2ClientId;
import cube8540.oauth.authentication.credentials.oauth.error.InvalidGrantException;
import cube8540.oauth.authentication.credentials.oauth.error.InvalidRequestException;
import cube8540.oauth.authentication.credentials.oauth.error.UserDeniedAuthorizationException;
import cube8540.oauth.authentication.credentials.oauth.token.domain.OAuth2AccessTokenRepository;
import cube8540.oauth.authentication.credentials.oauth.token.domain.OAuth2AuthorizedAccessToken;
import cube8540.oauth.authentication.credentials.oauth.token.domain.OAuth2TokenIdGenerator;
import cube8540.oauth.authentication.users.domain.UserEmail;
import org.springframework.security.authentication.AccountStatusException;
import org.springframework.security.authentication.AuthenticationManager;
import org.springframework.security.authentication.BadCredentialsException;
import org.springframework.security.authentication.UsernamePasswordAuthenticationToken;
import org.springframework.security.core.Authentication;
import org.springframework.security.oauth2.core.AuthorizationGrantType;

public class ResourceOwnerPasswordTokenGranter extends AbstractOAuth2TokenGranter {

    private final AuthenticationManager authenticationManager;

    public ResourceOwnerPasswordTokenGranter(OAuth2TokenIdGenerator tokenIdGenerator, OAuth2AccessTokenRepository tokenRepository,
                                             AuthenticationManager authenticationManager) {
        super(tokenIdGenerator, tokenRepository);
        this.authenticationManager = authenticationManager;
    }

    @Override
    public OAuth2AuthorizedAccessToken createAccessToken(OAuth2ClientDetails clientDetails, OAuth2TokenRequest tokenRequest) {
        if (tokenRequest.username() == null || tokenRequest.password() == null) {
            throw InvalidRequestException.invalidRequest("username, password is required");
        }
        if (!getTokenRequestValidator().validateScopes(clientDetails, tokenRequest.scopes())) {
            throw InvalidGrantException.invalidScope("cannot grant scopes");
        }
        Authentication authentication = authentication(tokenRequest);
        OAuth2AuthorizedAccessToken accessToken = OAuth2AuthorizedAccessToken.builder(getTokenIdGenerator())
                .expiration(extractTokenExpiration(clientDetails))
                .client(new OAuth2ClientId(clientDetails.clientId()))
                .email(new UserEmail(authentication.getName()))
                .scope(extractGrantScope(clientDetails, tokenRequest))
                .tokenGrantType(AuthorizationGrantType.PASSWORD)
                .build();
        accessToken.generateRefreshToken(refreshTokenGenerator(), extractRefreshTokenExpiration(clientDetails));
        return accessToken;
    }

    private Authentication authentication(OAuth2TokenRequest tokenRequest) {
        try {
            UsernamePasswordAuthenticationToken usernamePasswordToken =
                    new UsernamePasswordAuthenticationToken(tokenRequest.username(), tokenRequest.password());
            return authenticationManager.authenticate(usernamePasswordToken);
        } catch (BadCredentialsException | AccountStatusException e) {
            throw new UserDeniedAuthorizationException(e.getMessage());
        }
    }
}

File: src/main/kotlin/cube8540/oauth/authentication/oauth/token/application/ResourceOwnerPasswordTokenGranter.kt
package cube8540.oauth.authentication.oauth.token.application

import cube8540.oauth.authentication.oauth.client.domain.OAuth2ClientId
import cube8540.oauth.authentication.oauth.error.InvalidGrantException
import cube8540.oauth.authentication.oauth.error.InvalidRequestException
import cube8540.oauth.authentication.oauth.error.UserDeniedAuthorizationException
import cube8540.oauth.authentication.oauth.security.OAuth2ClientDetails
import cube8540.oauth.authentication.oauth.security.OAuth2TokenRequest
import cube8540.oauth.authentication.oauth.token.domain.OAuth2AccessTokenRepository
import cube8540.oauth.authentication.oauth.token.domain.OAuth2AuthorizedAccessToken
import cube8540.oauth.authentication.oauth.token.domain.OAuth2TokenIdGenerator
import cube8540.oauth.authentication.oauth.token.domain.PrincipalUsername
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.beans.factory.annotation.Qualifier
import org.springframework.security.authentication.AccountStatusException
import org.springframework.security.authentication.AuthenticationManager
import org.springframework.security.authentication.BadCredentialsException
import org.springframework.security.authentication.UsernamePasswordAuthenticationToken
import org.springframework.security.core.Authentication
import org.springframework.security.oauth2.core.AuthorizationGrantType
import org.springframework.stereotype.Service
import java.time.LocalDateTime

@Service
class ResourceOwnerPasswordTokenGranter @Autowired constructor(
    @Qualifier("defaultTokenIdGenerator")
    tokenIdGenerator: OAuth2TokenIdGenerator,

    tokenRepository: OAuth2AccessTokenRepository,

    @Qualifier("oauthAuthenticationBean")
    private val authenticationManager: AuthenticationManager
): AbstractOAuth2TokenGranter(tokenIdGenerator, tokenRepository) {

    override fun createAccessToken(clientDetails: OAuth2ClientDetails, tokenRequest: OAuth2TokenRequest): OAuth2AuthorizedAccessToken {
        if (tokenRequest.username == null || tokenRequest.password == null) {
            throw InvalidRequestException.invalidRequest("username, password is required")
        }
        if (!tokenRequestValidator.validateScopes(clientDetails, tokenRequest.scopes)) {
            throw InvalidGrantException.invalidScope("cannot grant scope")
        }

        val authentication = authentication(tokenRequest)
        val accessToken = OAuth2AuthorizedAccessToken(
            tokenIdGenerator = tokenIdGenerator,
            username = PrincipalUsername(authentication.name),
            client = OAuth2ClientId(clientDetails.clientId),
            scopes = extractGrantScope(clientDetails, tokenRequest).toMutableSet(),
            expiration = extractTokenExpiration(clientDetails),
            tokenGrantType = AuthorizationGrantType.PASSWORD,
            issuedAt = LocalDateTime.now(clock)
        )
        accessToken.generateRefreshToken(refreshTokenGenerator(), extractRefreshTokenExpiration(clientDetails))
        return accessToken
    }

    private fun authentication(tokenRequest: OAuth2TokenRequest): Authentication {
        try {
            val usernamePasswordToken = UsernamePasswordAuthenticationToken(tokenRequest.username, tokenRequest.password)
            return authenticationManager.authenticate(usernamePasswordToken)
        } catch (exception: Exception) {
            if (exception is BadCredentialsException || exception is AccountStatusException) {
                throw UserDeniedAuthorizationException(exception.message)
            } else {
                throw exception
            }
        }
    }

}